### PR TITLE
Detection and decoding of curved QR-codes

### PR DESCRIPTION
**Pull request by APrigarina**
These changes add detection and decoding of curved QR-codes

### CODE REVIEW

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
 protected:
     bool updatePerspective();
     bool versionDefinition();
     bool samplingForVersion();
     bool decodingProcess();
-    Mat original, no_border_intermediate, intermediate, straight;
+    inline double pointPosition(Point2f a, Point2f b , Point2f c);
+    float distancePointToLine(Point2f a, Point2f b , Point2f c);
+    bool getPointsInsideQRCode(vector<Point2f> angle_list);
+    bool computeClosestPoints(vector<Point> result_integer_hull);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:998]:
```suggestion
    bool computeClosestPoints(const vector<Point> &result_integer_hull);
```

File: modules/objdetect/src/qrcode.cpp [hunk 2]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    bool computeSidesPoints(vector<Point> result_integer_hull);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:999]:
```suggestion
    bool computeSidesPoints(const vector<Point> &result_integer_hull);
```

File: modules/objdetect/src/qrcode.cpp [hunk 3]
```diff
@@ -1001,206 +1073,1339 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+        else if (i == end - step)
+        {
+            enough_points = false;
+            index_max_angle = i;
         }
     }
+    
+    if (enough_points)
+    {
+        p1 = side.at(index_max_angle + step);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1262]: why do you use `at`?

File: modules/objdetect/src/qrcode.cpp [hunk 4]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
     bool versionDefinition();
     bool samplingForVersion();
     bool decodingProcess();
-    Mat original, no_border_intermediate, intermediate, straight;
+    inline double pointPosition(Point2f a, Point2f b , Point2f c);
+    float distancePointToLine(Point2f a, Point2f b , Point2f c);
+    bool getPointsInsideQRCode(vector<Point2f> angle_list);
+    bool computeClosestPoints(vector<Point> result_integer_hull);
+    bool computeSidesPoints(vector<Point> result_integer_hull);
+    vector<Point> getPointsNearUnstablePoint(vector<Point>&side, int start, int end, int step);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1000]:
```suggestion
    vector<Point> getPointsNearUnstablePoint(const vector<Point> &side, int start, int end, int step);
```

File: modules/objdetect/src/qrcode.cpp [hunk 5]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    bool findAndAddStablePoint(vector<Point> result_integer_hull);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1001]:
```suggestion
    bool findAndAddStablePoint(const vector<Point> &result_integer_hull);
```

File: modules/objdetect/src/qrcode.cpp [hunk 6]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    bool findIndexesCurvedSides();
+    bool findIncompleteIndexesCurvedSides();
+    Mat getMarkersMask();
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1004]: for more informatively 
```suggestion
    Mat getPatternsMask();
```

File: modules/objdetect/src/qrcode.cpp [hunk 7]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    Point findClosestZeroPoint(Point2f original_point);
+    bool findMarkersContours(vector<vector<Point>> &markers_contours);
+    bool findMarkersVerticesPoints(vector<vector<Point>> &markers_edge_points);
+    bool findTempMarkersAddingPoints(vector<std::pair<int,vector<Point>>> &temp_markers_add_points);
+    bool computeMarkersAddingPoints(std::map<int,vector<Point>> &markers_add_points);
+    bool addPointsToSides();
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1010]: for better measurement
```suggestion
    Point2f findClosestZeroPoint(Point2f original_point);
    bool findMarkersContours(vector<vector<Point2f> > &markers_contours);
    bool findMarkersVerticesPoints(vector<vector<Point2f> > &markers_edge_points);
    bool findTempMarkersAddingPoints(vector<std::pair<int, vector<Point2f> > > &temp_markers_add_points);
    bool computeMarkersAddingPoints(std::map<int, vector<Point2f> > &markers_add_points);
    bool addPointsToSides();
```

File: modules/objdetect/src/qrcode.cpp [hunk 8]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    vector<vector<float>> spline(vector<int> &x_arr, vector<int> &y_arr);
+    vector<vector<Point2f>> createSpline();
+    vector<vector<Point2f>> divideIntoEvenSegments();
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1013]:
```suggestion
    vector<vector<float> > computeSpline(vector<float> &x_arr, vector<float> &y_arr);
    vector<vector<Point2f> > createSpline();
    vector<vector<Point2f> > divideIntoEvenSegments();
```
and please replace `computeSpline` outside `QRDecode`

File: modules/objdetect/src/qrcode.cpp [hunk 9]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    bool straightenQRCodeInParts();
+    bool preparingCurvedQRCodes();
+    Mat original, bin_barcode, no_border_intermediate, intermediate, straight, curved_to_straight, test_image, hull_image, closest_points_image, sides_points_image, markers_image, spline_image, cut_points_image;
     vector<Point2f> original_points;
+    vector<Point2f> original_curved_points = vector<Point2f>(4);
+    vector<Point> qrcode_locations;
+    vector<std::pair<size_t,Point>> closest_points;
+    vector<vector<Point>> sides_points;
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1021]:
```suggestion
    vector<std::pair<size_t,Point> > closest_points;
    vector<vector<Point> > sides_points;
```

File: modules/objdetect/src/qrcode.cpp [hunk 10]
```diff
@@ -976,23 +986,85 @@ class QRDecode
...
+    std::pair<size_t,Point> unstable_pair;
+    bool approximate_unstable_point = false;
+    vector<int> curved_indexes, curved_incomplete_indexes;
+    std::map<int,vector<Point>> complete_curved_sides;
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1025]:
```suggestion
    std::map<int, vector<Point> > complete_curved_sides;
```

File: modules/objdetect/src/qrcode.cpp [hunk 11]
```diff
@@ -1001,212 +1069,1332 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+    return true;
+}
+
+bool QRDecode::computePatternsAddingPoints(std::map<int, vector<Point> > &patterns_add_points)
+{
+    vector<std::pair<int, vector<Point> > > temp_patterns_add_points;
+    if(!findTempPatternsAddingPoints(temp_patterns_add_points)) { return false; }
+    for(size_t i = 0; i < temp_patterns_add_points.size(); i++)
+    {
+        int num_points_in_pattern = 3;
```
Comment (rayonnant14) [inline modules/objdetect/src/qrcode.cpp:1696]:
```suggestion
        const int num_points_in_pattern = 3;
```
make all constant variables `const`

File: modules/objdetect/src/qrcode.cpp [hunk 12]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+        {
+            count++;
+            if (count == 3) continue;
+            double temp_norm = norm(it->second[j] -
+                                    it->second[j + 1]);
+            mean_step += temp_norm;
+        }
+        num_points_at_side += it->second.size();
+    }
+    mean_step /= num_points_at_side;
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1743]: please add test for check zero divided

File: modules/objdetect/src/qrcode.cpp [hunk 13]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+
+    std::map<int, vector<Point> >::iterator it;
+    double mean_step = 0.0;
+    size_t num_points_at_side = 0;
+    for (it=complete_curved_sides.begin(); it!=complete_curved_sides.end(); ++it)
+    {
+        int count = -1;
+        for(size_t j = 0; j < it->second.size() - 1; j++)
+        {
+            count++;
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1735]: please replace near with `j++`

File: modules/objdetect/src/qrcode.cpp [hunk 14]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+{
+    if(!computePatternsAddingPoints(complete_curved_sides)) { return false; }
+
+    std::map<int, vector<Point> >::iterator it;
+    double mean_step = 0.0;
+    size_t num_points_at_side = 0;
+    for (it=complete_curved_sides.begin(); it!=complete_curved_sides.end(); ++it)
+    {
+        int count = -1;
+        for(size_t j = 0; j < it->second.size() - 1; j++)
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1733]: please replace all constants outside cycles because cycle call `size()` function every iteration

File: modules/objdetect/src/qrcode.cpp [hunk 15]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+            {
+                complete_curved_sides[idx].push_back(sides_points[idx][sides_points_indexes[j]]);
+            }
+        }
+    }
+
+    return true;
+}
+
+vector<vector<float> > QRDecode::computeSpline(vector<int> &x_arr, vector<int> &y_arr)
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1789]:
```suggestion
vector<vector<float> > QRDecode::computeSpline(const vector<int> &x_arr, const vector<int> &y_arr)
```

File: modules/objdetect/src/qrcode.cpp [hunk 16]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+{
+    int n = (int)y_arr.size();
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1791]:
```suggestion
    const int n = (int)y_arr.size();
```

File: modules/objdetect/src/qrcode.cpp [hunk 17]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+    {
+        c[j] = z[j] - mu[j] * c[j + 1];
+        b[j] = (a[j + 1] - a[j]) / h[j] - (h[j] * (c[j + 1] + 2 * c[j])) / 3;
+        d[j] = (c[j + 1] - c[j]) / (3 * h[j]);
+    }
+
+    vector<vector<float> > S(n - 1);
+    for (int i = 0; i < n - 1; i++)
+    {
+        S[i].push_back(a[i]);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1831]: please use insert to add several elements

File: modules/objdetect/src/qrcode.cpp [hunk 18]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+
+        for (size_t j = 0; j < spline_points.size(); j++)
+        {
+            x_arr.push_back(cvRound(spline_points[j].x));
+            y_arr.push_back(cvRound(spline_points[j].y));
+        }
+        if (abs(x_arr.front() - x_arr.back()) > abs(y_arr.front() - y_arr.back()))
+        {
+            S = computeSpline(y_arr, x_arr);
+            if(closest_points[idx_curved_side].second.x >
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1860]: please use one condition branch

File: modules/objdetect/src/qrcode.cpp [hunk 19]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+            {
+                if (x == x_arr.front())
+                {
+                    spline_lines[idx].push_back(closest_points[start].second);
+                }
+                for (size_t i = 0; i < x_arr.size() - 1; i++)
+                {
+                    if ((x > x_arr[i]) && (x <= x_arr[i + 1]))
+                    {
+                        float y = S[i][0] + S[i][1] * (x - x_arr[i]) + S[i][2] * pow((x - x_arr[i]), 2)
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1881]: please use multiplication instead `pow`

File: modules/objdetect/src/qrcode.cpp [hunk 20]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+
+        vector<Point> spline_points = complete_curved_sides.find(idx_curved_side)->second;
+        vector<int> x_arr, y_arr;
+
+        for (size_t j = 0; j < spline_points.size(); j++)
+        {
+            x_arr.push_back(cvRound(spline_points[j].x));
+            y_arr.push_back(cvRound(spline_points[j].y));
+        }
+        if (abs(x_arr.front() - x_arr.back()) > abs(y_arr.front() - y_arr.back()))
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1857]: please use swap between `x` and `y` for calculate only one condition branch

File: modules/objdetect/src/qrcode.cpp [hunk 21]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+            }
+        }
+    }
+
+    return true;
+}
+
+bool QRDecode::divideIntoEvenSegments(vector<vector<Point2f> > &segments_points)
+{
+    vector<vector<Point2f> > spline_lines(NUM_SIDES);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1930]: please add `NUM_SIDES` into class

File: modules/objdetect/src/qrcode.cpp [hunk 22]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+    if (!createSpline(spline_lines)) { return false; }
+
+    float mean_num_points_in_line = 0.0;
+    for (size_t i = 0; i < NUM_SIDES; i++)
+    {
+        mean_num_points_in_line += spline_lines[i].size();
+    }
+    mean_num_points_in_line /= NUM_SIDES;
+    int min_num_points = 1, max_num_points = cvRound(mean_num_points_in_line / 2.0);
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1939]:
```suggestion
    const int min_num_points = 1, max_num_points = cvRound(mean_num_points_in_line / 2.0);
```

File: modules/objdetect/src/qrcode.cpp [hunk 23]
```diff
@@ -1001,214 +1069,1342 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+                        max_dist_to_line = temp_dist;
+                    }
+                }
+                mean_dist_in_segment += max_dist_to_line;
+            }
+            mean_dist_in_segment /= segments_points[i].size();
+            mean_of_two_sides    += mean_dist_in_segment;
+        }
+        mean_of_two_sides /= NUM_SIDES;
+        if (mean_of_two_sides < 0.5)
```
Comment (allnes) [inline modules/objdetect/src/qrcode.cpp:1984]: please name magic number

File: modules/objdetect/src/qrcode.cpp [hunk 24]
```diff
@@ -1001,216 +1070,1327 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+    version_size = 21 + (version - 1) * 4;
+    return true;
+}
+
+bool QRDecode::samplingForVersion()
+{
+    CV_TRACE_FUNCTION();
+    const double multiplyingFactor = (version < 3)  ? 1 :
+                                     (version == 3) ? 1.5 :
+                                     version * (5 + version - 4);
```
Comment (rayonnant14) [inline modules/objdetect/src/qrcode.cpp:2290]:
```suggestion
                                     version * (version + 1);
```

File: modules/objdetect/src/qrcode.cpp [hunk 25]
```diff
@@ -1001,217 +1070,1329 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+float QRDecode::distancePointToLine(Point2f a, Point2f b , Point2f c)
 {
-    // compute the width/height of convex hull
-    Rect areaBox = boundingRect(v);
+    float A, B, C, result;
+    A = c.y - b.y;
+    B = c.x - b.x;
+    C = c.x * b.y - b.x * c.y;
+    if (sqrt(A*A + B*B) == 0) return 0;
+    result = abs((A * a.x - B * a.y + C)) / sqrt(A*A + B*B);
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:1085]: Avoid calculation of `sqrt()` expression twice.

File: modules/objdetect/src/qrcode.cpp [hunk 26]
```diff
@@ -1001,217 +1070,1329 @@ void QRDecode::init(const Mat &src, const vector<Point2f> &points)
...
+        for(size_t j = 0; j < num_points_at_pattern - 1; j++, count++)
+        {
+            if (count == 3) continue;
+            double temp_norm = norm(it->second[j] -
+                                    it->second[j + 1]);
+            mean_step += temp_norm;
+        }
+        num_points_at_side += num_points_at_pattern;
+    }
+    if (num_points_at_side == 0) { return false; }
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp:1742]: Please use separate line for `if` body, even for simple cases.
It is important to see the real code coverage.

**Review by allnes — APPROVED** [review]

**Comment by ismaelbenito-ub** [comment]
Hi! Last year, I defended my PhD in Engineering and Applied Science, regarding a thesis which involved huge use of QR Codes, at the time we didn't realize OpenCV had included a curved QR Code decoder, it is amazing the work the community has been done in recent years. In the thesis, we tackled the recovery of QR Codes from curved surfaces using Thin Plate Splines (Chapter 3 of https://diposit.ub.edu/dspace/bitstream/2445/185845/1/IBA_PhD_THESIS.pdf). Is there any paper regarding the implementation here presented of the use of splines to extract the QR Code? If we understand correctly, the above presented method uses the splines to fit the outer contour of the QR Code? We would like to be able to compare our method and the one presented here, and if able to present the research of the comparison and later attempt to make a pull request. 

TL;DR: Looking for a paper related with this method.
